aquadoggo/src/worker.rs

```diff
@@ -0,0 +1,737 @@
+            let mut db = database.0.lock().map_err(|_| TaskError::Critical)?;
+
+            // 1. Identify unfinished puzzle related to this piece
+            let puzzle: Option<JigsawPuzzle> = db
+                .puzzles
+                .values()
+                .find(|item| item.piece_ids.contains(&input.id) && !item.complete)
+                .map(|item| item.clone());
+
+            // 2. Check if all piece dependencies are met
+            match puzzle {
+                None => Err(TaskError::Failure),
+                Some(mut puzzle) => {
+                    for piece_id in &puzzle.piece_ids {
+                        match db.pieces.get(&piece_id) {
+                            None => return Err(TaskError::Failure),
+                            Some(piece) => {
+                                for relation_piece_id in &piece.relations {
+                                    if !puzzle.piece_ids.contains(&relation_piece_id) {
+                                        return Err(TaskError::Failure);
+                                    }
+                                }
+                            }
+                        };
+                    }
+
+                    // Mark puzzle as complete! We are done here!
+                    puzzle.complete = true;
+                    db.puzzles.insert(puzzle.id, puzzle.clone());
+                    Ok(None)
+                }
+            }
+        }
+
+        // Register workers
+        factory.register("pick", 3, pick);
+        factory.register("find", 3, find);
+        factory.register("finish", 3, finish);
+
+        // Generate a number of puzzles to solve
+        let puzzles_count = 10;
+        let min_size = 3;
+        let max_size = 10;
+
+        let mut pieces: Vec<JigsawPiece> = Vec::new();
+        let mut offset: isize = 0;
+
+        for _ in 0..puzzles_count {
+            let size = rand::thread_rng().gen_range(min_size..max_size);
+            let mut id: isize = 0;
+
+            for _ in 0..size {
+                for _ in 0..size {
+                    let mut relations: Vec<usize> = Vec::new();
+
+                    id = id + 1;
+
+                    if id % size != 0 {
+                        relations.push((offset + id + 1) as usize);
+                    }
+
+                    if id % size != 1 {
+                        relations.push((offset + id - 1) as usize);
+                    }
+
+                    if id + size <= size * size {
+                        relations.push((offset + id + size) as usize);
+                    }
+
+                    if id - size > 0 {
+                        relations.push((offset + id - size) as usize);
+                    }
+
+                    pieces.push(JigsawPiece {
+                        id: (offset + id) as usize,
+                        relations,
+                    });
+                }
+            }
+
+            offset = offset + (size * size);
+        }
```

Comment:
I can't quite grok what's happening here... can you expand a little?

Reply:
I've added some comments, does that help?

Reply:
Yes, perfect! Thanks.